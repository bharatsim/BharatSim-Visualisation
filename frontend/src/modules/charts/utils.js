import { chartStyleConfig, getColor } from './chartStyleConfig';

function trasformDataForChart(csvData, xColumn, yColumns) {
  return {
    labels: csvData && csvData.data[xColumn],
    datasets:
      csvData &&
      yColumns.map((yColumn, index) => {
        return {
          ...chartStyleConfig,
          label: yColumn,
          borderColor: getColor(index),
          backgroundColor: getColor(index),
          data: csvData && csvData.data[yColumn],
        };
      }),
  };
}

function getYaxisNames(yColumns) {
  return yColumns.map((yColumn) => yColumn.name);
}

export { trasformDataForChart, getYaxisNames };
