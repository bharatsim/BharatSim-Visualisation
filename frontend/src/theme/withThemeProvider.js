import React from 'react';
import { MuiThemeProvider, CssBaseline } from '@material-ui/core';

import theme from './theme';

function withThemeProvider(WrappedComponent) {
  return function WithTheme(props) {
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <WrappedComponent {...props} />
      </MuiThemeProvider>
    );
  };
}

export default withThemeProvider;
