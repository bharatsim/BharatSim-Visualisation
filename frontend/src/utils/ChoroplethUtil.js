import L from 'leaflet';
import chroma from 'chroma-js';

function highlightOnClick(e) {
  const layer = e.target;

  layer.setStyle({
    weight: 2,
    dashArray: '',
    opacity: 1,
    fillOpacity: 0.7,
  });

  if (!L.Browser.ie && !L.Browser.opera && !L.Browser.edge) {
    layer.bringToFront();
  }
}

function resetHighlight(geoJson) {
  return (e) => geoJson.current.resetStyle(e.target);
}

function zoomToFeature(map) {
  return (e) => map.fitBounds(e.target.getBounds());
}

function geoJSONStyle(idDataMap, mapLayerIdName, maxOfMeasure, scale) {
  const colors = Object.keys(scale)
    .sort((v1, v2) => Number(v1) - Number(v2))
    .map((key) => scale[key]);
  const getColor = chroma.scale(colors);
  return (feature) => {
    const gradientPercentage = idDataMap[feature.properties[mapLayerIdName]] / maxOfMeasure;
    return {
      color: '#1f2021',
      weight: 1,
      fillOpacity: 0.5,
      opacity: 0.7,
      fillColor: getColor(gradientPercentage),
    };
  };
}

export { highlightOnClick, resetHighlight, geoJSONStyle, zoomToFeature };
