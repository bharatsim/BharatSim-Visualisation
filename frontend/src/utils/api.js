import { contentTypes, httpMethods } from '../constants/fetch';
import { url as serviceURL } from './url';
import { fetchData, uploadData } from './fetch';

function headerBuilder({ contentType }) {
  return {
    'content-type': contentType,
  };
}

function formDataBuilder(data) {
  const formData = new FormData();
  data.forEach(({ name, value }) => formData.append(name, value));
  return formData;
}

const api = {
  saveDashboard: async ({ widgets, layout, dashboardId, name, count }) => {
    return uploadData({
      url: serviceURL.DASHBOARD_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets, layout, dashboardId, name, count },
      }),
    });
  },

  addNewDashboard: async ({ name, projectId }) => {
    return uploadData({
      url: serviceURL.INSERT_DASHBOARD,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets: [], layout: [], name, count: 0, projectId },
      }),
      isCustomErrorHandler: true,
    });
  },

  getAllDashBoardByProjectId: async (projectId) =>
    fetchData({
      url: serviceURL.DASHBOARD_URL,
      query: { projectId, columns: ['name', '_id'] },
    }),

  getAllDashBoard: async () => fetchData({ url: serviceURL.DASHBOARD_URL }),

  uploadFileAndSchema: async ({ file, schema, dashboardId }) =>
    uploadData({
      url: serviceURL.DATA_SOURCES,
      data: formDataBuilder([
        { name: 'datafile', value: file },
        { name: 'schema', value: JSON.stringify(schema) },
        { name: 'dashboardId', value: dashboardId },
      ]),
      headers: headerBuilder({ contentType: contentTypes.FILE }),
    }),

  getCsvHeaders: async (dataSourceId) => fetchData({ url: serviceURL.getHeaderUrl(dataSourceId) }),

  getDatasources: async (dashboardId) =>
    fetchData({
      url: serviceURL.DATA_SOURCES,
      query: { dashboardId },
    }),

  getData: async (datasource, columns) =>
    fetchData({ url: serviceURL.getDataUrl(datasource), query: { columns } }),

  saveProject: async ({ id, ...data }) => {
    const requestObject = {
      url: serviceURL.PROJECT_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      isCustomErrorHandler: true,
    };
    if (id) {
      return uploadData({
        ...requestObject,
        data: JSON.stringify({ projectData: { ...data, id } }),
        method: httpMethods.PUT,
      });
    }
    return uploadData({
      ...requestObject,
      data: JSON.stringify({ projectData: { ...data } }),
      method: httpMethods.POST,
    });
  },

  getProjects: async () => {
    return fetchData({
      url: serviceURL.PROJECT_URL,
    });
  },
  getProject: async (id) => {
    return fetchData({
      url: serviceURL.getProjectUrl(id),
    });
  },
};

export { api };
