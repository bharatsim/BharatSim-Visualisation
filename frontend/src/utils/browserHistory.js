// eslint-disable-next-line import/no-mutable-exports
let BrowserHistory;

function initHistory(history) {
  BrowserHistory = history;
}

function navigateToPreviousPage() {
  BrowserHistory.goBack();
}

function navigateToHome() {
  BrowserHistory.push('/');
}

function reloadPage() {
  // eslint-disable-next-line no-restricted-globals
  location.reload();
}

export { initHistory, navigateToPreviousPage, reloadPage, BrowserHistory, navigateToHome };
