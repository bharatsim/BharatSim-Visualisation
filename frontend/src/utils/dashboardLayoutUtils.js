export function getNewWidgetLayout(numberOfWidgetAdded, cols, count) {
  return createLayout({
    id: `widget-${count}`,
    xPosition: (numberOfWidgetAdded * 6) % cols,
    yPosition: Infinity, // puts it at the bottom
    width: 6,
    height: 2,
  });
}

function createLayout({ id, xPosition, yPosition, width, height }) {
  return {
    i: id,
    x: xPosition,
    y: yPosition,
    w: width,
    h: height,
  };
}
