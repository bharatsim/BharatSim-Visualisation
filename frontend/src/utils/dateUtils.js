const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

function getPeriod(hours) {
  return hours >= 12 ? 'PM' : 'AM';
}

function formatDate(dateString) {
  const date = new Date(dateString);
  if (date.toString() === 'Invalid Date') return '--';
  const DD = date.getDate();
  const MM = date.getMonth();
  const YYYY = date.getFullYear();
  const HH = date.getHours();
  const mm = date.getMinutes();

  return `${DD}-${months[MM]}-${YYYY} at ${HH % 12}:${mm < 10 ? `0${mm}` : mm} ${getPeriod(HH)}`;
}

function currentDate(){
  return new Date();
}

export { formatDate, currentDate };
