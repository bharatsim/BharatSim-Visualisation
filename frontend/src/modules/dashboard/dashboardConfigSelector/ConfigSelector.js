import React, { useEffect, useState } from 'react';
import Box from '@material-ui/core/Box';
import Divider from '@material-ui/core/Divider';
import { useForm } from 'react-final-form';

import { useSnackbar } from 'notistack';
import chartConfigs from '../../../config/chartConfigs';
import { api } from '../../../utils/api';
import chartConfigOptions from '../../../config/chartConfigOptions';
import useLoader from '../../../hook/useLoader';
import LoaderOrError from '../../loaderOrError/LoaderOrError';
import { chartConfigOptionTypes } from '../../../constants/chartConfigOptionTypes';
import { useFormContext } from '../../../contexts/FormContext';
import { compareArrayByValues } from '../../../utils/helper';
import snackbarVariant from '../../../constants/snackbarVariant';

function isMatchingHeaders(newHeaders, prevHeaders) {
  const newColumns = newHeaders.headers.map(({ name }) => name);
  const prevColumns = prevHeaders.headers.map(({ name }) => name);
  return compareArrayByValues(newColumns, prevColumns);
}

function ConfigSelector() {
  const [fetchedCsvHeaders, setFetchedCsvHeaders] = useState();
  const { chartType } = useFormContext();
  const { getFieldState, reset } = useForm();
  const { enqueueSnackbar } = useSnackbar();

  const dataSourceId = getFieldState(chartConfigOptionTypes.DATASOURCE)?.value;

  const {
    startLoader,
    stopLoaderAfterSuccess,
    stopLoaderAfterError,
    loadingState,
    message,
  } = useLoader();

  useEffect(() => {
    if (dataSourceId) fetchCsvHeaders();
  }, [dataSourceId]);

  function resetConfig() {
    const chartName = getFieldState(chartConfigOptionTypes.CHART_NAME)?.value;
    reset({
      [chartConfigOptionTypes.DATASOURCE]: dataSourceId,
      [chartConfigOptionTypes.CHART_NAME]: chartName,
    });
  }

  function onFetchCsvHeaderSuccess(resData) {
    const shouldResetConfig = fetchedCsvHeaders && !isMatchingHeaders(resData, fetchedCsvHeaders);
    if (shouldResetConfig) {
      resetConfig();
      enqueueSnackbar('Configuration wizard form got reset for new data-source selection', {
        variant: snackbarVariant.INFO,
      });
    }
    stopLoaderAfterSuccess();
    setFetchedCsvHeaders(resData);
  }

  async function fetchCsvHeaders() {
    startLoader();
    api
      .getCsvHeaders(dataSourceId)
      .then(onFetchCsvHeaderSuccess)
      .catch(() => {
        stopLoaderAfterError('Unable to fetch data source headers');
      });
  }

  const { headers } = fetchedCsvHeaders || {};
  const chartConfigProps = { headers };
  const configOptionsKeysForSelectedChart = chartConfigs[chartType].configOptions;
  function isLastConfigOption(configs, index) {
    return configs.length - index === 1;
  }

  const onErrorAction = {
    name: 'Retry',
    onClick: fetchCsvHeaders,
  };

  return (
    <LoaderOrError
      loadingState={loadingState}
      message={message}
      errorAction={onErrorAction}
      fullWidth
    >
      <>
        {headers && (
          <div>
            {configOptionsKeysForSelectedChart.map((chartConfigKey, index) => (
              <Box key={chartConfigKey} pb={4}>
                <Box pb={4}>{chartConfigOptions[chartConfigKey].component(chartConfigProps)}</Box>
                {isLastConfigOption(configOptionsKeysForSelectedChart, index) ? '' : <Divider />}
              </Box>
            ))}
          </div>
        )}
      </>
    </LoaderOrError>
  );
}

export default ConfigSelector;
