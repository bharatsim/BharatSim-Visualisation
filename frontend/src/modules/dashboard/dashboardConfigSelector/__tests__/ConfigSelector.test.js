import { render } from '@testing-library/react';
import React, { useEffect } from 'react';
import { FormProvider, useForm } from 'react-hook-form';
import { fireEvent } from '@testing-library/dom';
import ConfigSelector from '../ConfigSelector';
import withThemeProvider from '../../../../theme/withThemeProvider';
import { api } from '../../../../utils/api';

jest.mock('../../../../utils/api', () => ({
  api: {
    getCsvHeaders: jest.fn().mockResolvedValue({
      headers: [
        { name: 'column1', type: 'number' },
        { name: 'column2', type: 'number' },
      ],
    }),
  },
}));

jest.mock('../../../../config/chartConfigOptions', () => ({
  xAxis: {
    component: () => <div>select x axis</div>,
  },
  yAxis: {
    component: () => <div>select y axis</div>,
  },
}));

const ConfigSelectorWithProvider = withThemeProvider(ConfigSelector);

const TestForConfigSelector = ({ onSubmit, isEditMode, chartType, resetValue, dataSourceId }) => {
  const form = useForm({ mode: 'onChange', defaultValues: { dataSource: dataSourceId } });
  const { handleSubmit, reset } = form;

  useEffect(() => {
    reset({ dataSource: dataSourceId });
  }, [dataSourceId]);

  const methods = {
    ...form,
    defaultValues: { dataSource: dataSourceId },
    isEditMode,
    resetValue,
    chartType,
  };
  return (
    <FormProvider {...methods}>
      <form onSubmit={handleSubmit(onSubmit)}>
        <ConfigSelectorWithProvider />
        <button type="submit">submit</button>
      </form>
    </FormProvider>
  );
};

describe('<ConfigSelector />', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot for configs for line chart', async () => {
    const { container, findByText } = render(
      <TestForConfigSelector
        dataSourceId="datasourceID"
        chartType="lineChart"
        resetValue={jest.fn()}
        isEditMode={false}
      />,
    );

    await findByText('select x axis');

    expect(container).toMatchSnapshot();
  });

  it('should call getCsvHeaders with data source id on render', async () => {
    const renderedComponent = render(
      <TestForConfigSelector
        dataSourceId="datasourceID"
        chartType="lineChart"
        resetValue={jest.fn()}
        errors={{}}
        isEditMode={false}
        control={{}}
        register={jest.fn()}
        watch={jest.fn()}
      />,
    );
    await renderedComponent.findByText('select x axis');

    expect(api.getCsvHeaders).toHaveBeenCalledWith('datasourceID');
  });

  it('should call getCsvHeaders with data source id on rerender for data source id change', async () => {
    const { findByText, rerender } = render(
      <TestForConfigSelector
        dataSourceId="datasourceID"
        chartType="lineChart"
        resetValue={jest.fn()}
        isEditMode={false}
      />,
    );

    await findByText('select x axis');

    rerender(
      <TestForConfigSelector
        dataSourceId="datasourceID2"
        chartType="lineChart"
        resetValue={jest.fn()}
        isEditMode={false}
      />,
    );

    await findByText('select x axis');

    expect(api.getCsvHeaders).toHaveBeenLastCalledWith('datasourceID2');
  });

  it('should call resetValue for config of line chart', async () => {
    const resetValue = jest.fn();
    const { findByText, rerender } = render(
      <TestForConfigSelector
        dataSourceId="datasourceID"
        chartType="lineChart"
        resetValue={resetValue}
        errors={{}}
        isEditMode={false}
        control={{}}
        register={jest.fn()}
        watch={jest.fn()}
      />,
    );

    await findByText('select x axis');

    rerender(
      <TestForConfigSelector
        dataSourceId="datasourceID2"
        chartType="lineChart"
        resetValue={jest.fn()}
        errors={{}}
        isEditMode={false}
        control={{}}
        register={jest.fn()}
        watch={jest.fn()}
      />,
    );

    await findByText('select x axis');

    expect(resetValue).toHaveBeenCalledWith(['xAxis', 'yAxis']);
  });

  it('should show loader while fetching data', async () => {
    const renderedComponent = render(
      <TestForConfigSelector
        dataSourceId="datasourceID"
        chartType="lineChart"
        resetValue={jest.fn()}
        errors={{}}
        isEditMode={false}
        control={{}}
        register={jest.fn()}
        watch={jest.fn()}
      />,
    );
    const loaderComponent = document.getElementsByTagName('svg');

    expect(loaderComponent).not.toBeNull();

    await renderedComponent.findByText('select x axis');
  });

  it('should show error if error occur while fetching data', async () => {
    api.getCsvHeaders.mockRejectedValueOnce('error');
    const { findByText, getByText } = render(
      <TestForConfigSelector
        dataSourceId="datasourceID"
        chartType="lineChart"
        resetValue={jest.fn()}
        errors={{}}
        isEditMode={false}
        control={{}}
        register={jest.fn()}
        watch={jest.fn()}
      />,
    );

    await findByText('Unable to fetch data source headers');

    expect(getByText('Unable to fetch data source headers')).toBeInTheDocument();
  });

  it('should refetch data on click on retry button present on error banner', async () => {
    api.getCsvHeaders.mockRejectedValueOnce('error');
    const { findByText, getByText } = render(
      <TestForConfigSelector
        dataSourceId="datasourceID"
        chartType="lineChart"
        resetValue={jest.fn()}
        errors={{}}
        isEditMode={false}
        control={{}}
        register={jest.fn()}
        watch={jest.fn()}
      />,
    );

    await findByText('Unable to fetch data source headers');
    const retryButton = getByText('Retry').closest('button');
    fireEvent.click(retryButton);

    await findByText('select x axis');

    expect(api.getCsvHeaders).toHaveBeenCalled();
  });
});
