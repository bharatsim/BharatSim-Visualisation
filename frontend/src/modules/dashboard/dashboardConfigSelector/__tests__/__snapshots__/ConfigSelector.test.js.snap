// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConfigSelector /> should match snapshot for configs for line chart 1`] = `
<div>
  <form>
    <div>
      <div
        class="MuiBox-root MuiBox-root-3"
      >
        <div
          class="MuiBox-root MuiBox-root-4"
        >
          <div>
            select x axis
          </div>
        </div>
        <hr
          class="MuiDivider-root"
        />
      </div>
      <div
        class="MuiBox-root MuiBox-root-5"
      >
        <div
          class="MuiBox-root MuiBox-root-6"
        >
          <div>
            select y axis
          </div>
        </div>
        <hr
          class="MuiDivider-root"
        />
      </div>
      <div
        class="MuiBox-root MuiBox-root-7"
      >
        <div
          class="MuiBox-root MuiBox-root-8"
        >
          <div>
            annotation
          </div>
        </div>
        
      </div>
    </div>
    <button
      type="submit"
    >
      submit
    </button>
  </form>
</div>
`;
